src/java/org/apache/cassandra/config/CassandraRelevantProperties.java

```diff
@@ -352,14 +410,166 @@
      * when the JVM terminates. Therefore, we can use such optimization and not wait unnecessarily. */
     NON_GRACEFUL_SHUTDOWN("cassandra.test.messagingService.nonGracefulShutdown"),
+    CHRONICLE_ANNOUNCER_DISABLE("chronicle.announcer.disable"),
+    COMMITLOG_ALLOW_IGNORE_SYNC_CRC("cassandra.commitlog.allow_ignore_sync_crc"),
+    COMMITLOG_IGNORE_REPLAY_ERRORS("cassandra.commitlog.ignorereplayerrors"),
+    COMMITLOG_MAX_OUTSTANDING_REPLAY_BYTES("cassandra.commitlog_max_outstanding_replay_bytes", String.valueOf(1024 * 1024 * 64)),
+    COMMITLOG_MAX_OUTSTANDING_REPLAY_COUNT("cassandra.commitlog_max_outstanding_replay_count", "1024"),
+    COMMITLOG_STOP_ON_ERRORS("cassandra.commitlog.stop_on_errors"),
+    CONFIG_LOADER("cassandra.config.loader"),
+    CONSISTENT_RANGE_MOVEMENT("cassandra.consistent.rangemovement", "true"),
+    CONSISTENT_SIMULTANEOUS_MOVES_ALLOW("cassandra.consistent.simultaneousmoves.allow", "false"),
+    CUSTOM_GUARDRAILS_CONFIG_PROVIDER_CLASS("cassandra.custom_guardrails_config_provider_class"),
+    CUSTOM_QUERY_HANDLER_CLASS("cassandra.custom_query_handler_class"),
+    CUSTOM_TRACING_CLASS("cassandra.custom_tracing_class"),
+    DATA_OUTPUT_STREAM_PLUS_TEMP_BUFFER_SIZE("cassandra.data_output_stream_plus_temp_buffer_size", "8192"),
+    DECAYING_ESTIMATED_HISTOGRAM_RESERVOIR_STRIPE_COUNT("cassandra.dehr_stripe_count", "2"),
+    DIAGNOSTIC_SNAPSHOT_INTERVAL_NANOS("cassandra.diagnostic_snapshot_interval_nanos", "60000000000"),
+    DISABLE_AUTH_CACHES_REMOTE_CONFIGURATION("cassandra.disable_auth_caches_remote_configuration"),
+    DISABLE_PAXOS_AUTO_REPAIRS("cassandra.disable_paxos_auto_repairs"),
+    DISABLE_PAXOS_STATE_FLUSH("cassandra.disable_paxos_state_flush"),
+    DISABLE_STCS_IN_L0("cassandra.disable_stcs_in_l0"),
+    DISABLE_TCACTIVE_OPENSSL("cassandra.disable_tcactive_openssl"),
+    DOB_DOUBLING_THRESHOLD_MB("cassandra.DOB_DOUBLING_THRESHOLD_MB", "64"),
+    DOB_MAX_RECYCLE_BYTES("cassandra.dob_max_recycle_bytes", String.valueOf(1024 * 1024)),
+    DROP_OVERSIZED_READ_REPAIR_MUTATIONS("cassandra.drop_oversized_readrepair_mutations"),
+    DTEST_API_LOG_TOPOLOGY("cassandra.dtest.api.log.topology"),
+    ENABLE_DC_LOCAL_COMMIT("cassandra.enable_dc_local_commit", "true"),
+    EXPIRATION_DATE_OVERFLOW_POLICY("cassandra.expiration_date_overflow_policy"),
+    EXPIRATION_OVERFLOW_WARNING_INTERVAL_MINUTES("cassandra.expiration_overflow_warning_interval_minutes", "5"),
+    FD_INITIAL_VALUE_MS("cassandra.fd_initial_value_ms"),
+    FD_MAX_INTERVAL_MS("cassandra.fd_max_interval_ms"),
+    FILE_CACHE_ENABLED("cassandra.file_cache_enabled"),
+    FORCE_DEFAULT_INDEXING_PAGE_SIZE("cassandra.force_default_indexing_page_size"),
+    FORCE_PAXOS_STATE_REBUILD("cassandra.force_paxos_state_rebuild"),
+    GIT_SHA("cassandra.gitSHA"),
+    GOSSIP_DISABLE_THREAD_VALIDATION("cassandra.gossip.disable_thread_validation", "false"),
+    IGNORE_CORRUPTED_SCHEMA_TABLES("cassandra.ignore_corrupted_schema_tables", "false"),
```

Comment:
This is interesting ... 

    public boolean getBoolean()
    {
        String value = System.getProperty(key);

        return BOOLEAN_CONVERTER.convert(value == null ? defaultVal : value);
    }

So when `defaultValue` is null, and `value` is null, then `BOOLEAN_CONVERTER.convert` will be called with `null`.

Then 

    PropertyConverter<Boolean> BOOLEAN_CONVERTER = Boolean::parseBoolean;

So finally, 

    public static boolean parseBoolean(String s) {
        return ((s != null) && s.equalsIgnoreCase("true"));
    }

So if no default is used in case of boolean, it will default it to "false" so we may probably omit that "false" in these cases.

Reply:
Fixed.

Reply:
@smiklosovic , @jacek-lewandowski 

Folks, I'd like to continue with this as I've found some inconsistencies in the source code regarding the use of defaults. I'd like to discuss with you some thoughts about the `CassandraRelevantProperties` default values. My concern here is that the current API for getting the value of system properties doesn't give us a reliable and explicit way to get the same results each time we access a property, so removing "redundant" defaults may cause more problems than adding them explicitly. 

### Current state

Let's take a look at a few examples and inconsistencies:

1. Inconsistency between `getBoolean()` and `getInt()`/`getLong()` methods:
* in case `getBoolean()` without defaults value the `false` value returned, no exception is thrown;
* in case `getInt()` or `getLong()` with default value the exception is thrown (`NullPointerException` :-( see [testInteger_null()](https://github.com/apache/cassandra/blob/trunk/test/unit/org/apache/cassandra/config/CassandraRelevantPropertiesTest.java#L133) test);
2. Inconsistency between `getBoolean()` and `getString()` for the same property (assume a system property with type boolean is used):
* `getBoolean()` will return a non-null value if the default value is `null`;
* `getString()` for the same property will return a `null` value;
3. If the default value is `null` the `NullPointerException` is thrown:
The `getBoolean()` method implicitly returns the `false` value and depends on the boolean converter implementation itself (so a developer has to delve into the code to find the _real_ return value). However, the `public long getLong()` will throw `NullPointerException`


### Expected state

So, I think the API should be the following (default value `null` is allowed) and we have to fix it:

1. `boolean disableSTCSInL0 = DISABLE_STCS_IN_L0.getBoolean(false);`
- If the property is set, the corresponding system property value is used;
- If the property is NOT set, the default value from brackets is used;

_OR_

2. `DISABLE_STCS_IN_L0("cassandra.disable_stcs_in_l0", "false")`
- If the property is set, the corresponding system property value is used;
- If the property is NOT set, the default value from the constant is used;

_OR_

3. `boolean disableSTCSInL0 = DISABLE_STCS_IN_L0.getBoolean()`
- If the property is set, the corresponding system property value is used;
- If the property is NOT set, and there is NO default value we MUST throw an exception.


WDYT?

Reply:
The expected behaviour for the `getBoolean()` (or similar methods) is to throw a `ConfigurationException` if neither the explicitly provided default value e.g. `DISABLE_STCS_IN_L0.getBoolean(false)` or no default value is set for the property e.g. `DISABLE_STCS_IN_L0("cassandra.disable_stcs_in_l0")`.

Here is an example:
https://github.com/Mmuzaf/cassandra/blob/cassandra-17797-default/src/java/org/apache/cassandra/config/CassandraRelevantProperties.java#L552

Reply:
@jacek-lewandowski 

If we're talking about the API for Cassandra's properties, it's not enough to just focus on the `CassandraRelevantProperties` class itself, and I'd say we should look at the scope a bit wider and put all configuration properties (configuration, system properties or whatever) under the same kind of interfaces and API. It was partially mentioned in Benedict's comment [here](https://issues.apache.org/jira/browse/CASSANDRA-15254?focusedCommentId=17451067&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17451067) and I have also been confronted with a lack of API to access config properties to expose them through the settings virtual table - [here](https://github.com/apache/cassandra/pull/2133/files#diff-931d44ec5e102506bca140864bb516f55c00028a95392b7eee3c647f22293339R56).

I'm working on the API to access the configuration properties for virtual tables (the `getBoolean`, `getInt` methods for primitive values are omitted in the draft implementation) - the `Registry` prototyped [here](https://github.com/apache/cassandra/pull/2133/files#diff-58bd5ac9ec3b4e19a95f2669860fa72d69f4290354376662e8d8772f64af05feR37), and I think it will be good to reuse the API for all sorts of configuration property stores in Cassandra, and also align the property naming in the end. You've also partially mentioned that we should have the same interface for CassandraRelevantProperties and CassandraRelevantEnv, which I agree with.

So I think we are talking about the same thing from slightly different angles. We should have the `Registry` - a common API interface for accessing properties with:
- `ConfiguraitonRegistry` - to access the properties stored in the Config class;
- `CassandraRelevantProperties` - to access the Cassandra's system properties;
- `CassandraRelevantEnv` - to access the Cassandra's environment variables;

Having everything above under the common interface allows us to build a configuration model tree for all of the ways Cassandra is configured.

## The API in Design Details

I think we should have a discussion on the dev-list for sure, but as a draft, I can share the following:

1. I think we should allow default values for Cassandra's system properties and forbid the same for non-cassandra properties as you mentioned it (e.g. for those that are not followed with `cassandra.` forbid defaults). We have defaults for some of the properties in the Config class, so it sounds logical to keep the same behaviour for the `CassandraRelevantProperties`.

2. `getBooleanOpt()` I'm not against this method at all, but the `Optional` is not enough for all the cases we have. We should allow internally in the `map` method to handle not only `unchecked` exceptions, but also checked ones that may affect the final implementation. The example is [here](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/config/DatabaseDescriptor.java#L1839).

3. If we go with triplets like `getBoolean()`, `getBoolean(boolean defaultValue)`, `getBooleanOpt()`, we will have them for every primitive type used by `CassandraRelevantProperties` variables. So the final API will have the 3 times multiplier on the total methods and considering the fact that Cassandra's `getBooleanOpt()` should also be aware of checked exceptions, the complexity may become even greater.

My guess is that we could have an intermediate interface for getting values. I have the following `Registry` [here](https://github.com/apache/cassandra/pull/2133/files#diff-58bd5ac9ec3b4e19a95f2669860fa72d69f4290354376662e8d8772f64af05feR37), please note it doesn't take into account the `Optional` approach you suggested, but if we want to go with `Optional` we could have something similar:
```
public interface Registry
{
  ConfigValue get(String key);
}
```

```
public interface ConfigValue
{
  Object get();
  String convertToString();
  boolean getBoolean();
  int getInt();
  long getLong();

  Class<?> type();
  ConfigValue map(CheckedFunction<? super ConfigValue, ? extends ConfigValue> mapper);
  boolean ifPresent();
}
```

4. If we go with the changes above I think we should add type for each property in the `ConfiguraitonRegistry` e.g. `ACQUIRE_RETRY_SECONDS("cassandra.acquire_retry_seconds", Integer.class, "60")`


## Current changes

Going back to the scope of these changes, I can propose the following plan:

- forbid the defaults for non-cassandra system properties in this PR;
- `getBoolean()` should throw an exception if the value does not exist;
- set system property defaults or use `getBoolean(bolean overrideDefaultValue)` where it's possible;
- The combination of `ifPresent()`, `getBoolean()`, `getBoolean(boolean overrideDefaultValue)` covers all the cases we have now;
- Create an API design issue for all Config, CassandraRelevantProperties, CassandraRelevantEnv classes or do the design part under [CASSANDRA-15254](https://issues.apache.org/jira/browse/CASSANDRA-15254) as the `Registry` is one of the requirements for closing that issue.


Please, share your thoughts on this plan.

Reply:
I really think we should deliver changes mentioned there under other tickets. Lets really just focus on the housekeeping here and centralizing it into CRP. After that we may do other changes. There is nothing wrong if work has some iterations, even more so if it is of such scope. I think that the patch itself is already big enough and we are starting to go too deep into it without actually delivering anything.

Reply:
@jacek-lewandowski @smiklosovic 

I have prototyped a configuration approach you suggested and it looks good at first glance. So, we could have multiple configuration sources under the same interface:
- `Config` class;
- `CassandraRelevantProperties`
- `CassandraRelevantEnv`

Having at the same time a read-only access layer represented with `ConfigurationRegistry` + `ConfgurtionValue<T>`, we will be able to turn the following source code:

```
if (conf.auto_snapshot_ttl != null)
{
    try
    {
        holder.autoSnapshoTtl = new DurationSpec.IntSecondsBound(conf.auto_snapshot_ttl);
    }
    catch (IllegalArgumentException e)
    {
        throw new ConfigurationException("Invalid value of auto_snapshot_ttl: " + conf.auto_snapshot_ttl, false);
    }
}
```

into

```
registry.getValue(String.class, "auto_snapshot_ttl")
        .map(DurationSpec.IntSecondsBound::new,
             (e, v) -> new ConfigurationException("Invalid value of auto_snapshot_ttl: " + v, false))
        .ifPresent(v -> holder.autoSnapshoTtl = v);
```

It looks much better and suits our needs. Here is the prototype of the `ConfigurationValue<T>` class:
https://github.com/apache/cassandra/compare/trunk...Mmuzaf:cassandra:cassandra-15254-apply#diff-16d8aa644f29a8e76deedbb8dd760530530c90d4dd14ad4c4ea60823b84ea79dR31


I guess we could implement that in another issue (cassandra-15254?), so here we just need to decide what to do with the CRP default values.
WDYT?

Reply:
@Mmuzaf as Jacek suggested lets just gather all these improvements in an epic in Jira or similar ....